src/main/kotlin/io/github/inductiveautomation/kindling/core/PreferencesPanel.kt

```diff
@@ -0,0 +1,173 @@
+                    add(cancelButton, "east, gapright 10")
+                    add(submitButton, "east, gapright 10")
+                },
+                "south, gapbottom 10"
+            )
+        }
+        pack()
+    }
+
+    private fun saveChanges() {
+        Kindling.session.showFullLoggerNames = generalPanel.showFullLoggerNames.isSelected
+        Kindling.session.saveAndResume = generalPanel.sessionResume.isSelected
+        Kindling.session.uiScaleFactor = uiPanel.uiScaleSelector.value as Double
+
+        if (Kindling.session.theme != uiPanel.themeSelection.selectedItem) {
+            Kindling.session.theme = uiPanel.themeSelection.selectedItem
+        }
+    }
+
+    class UIPanel : JPanel(MigLayout("fill, gap 10")) {
+        private val lightDarkToggle = JToggleButton().apply {
+            if (Kindling.session.theme.isDark) {
+                isSelected = false
+                text = "Dark Themes"
+            } else {
+                isSelected = true
+                text = "Light Themes"
+            }
+
+            addItemListener { event ->
+                text = if (event.stateChange == ItemEvent.SELECTED) "Light Themes" else "Dark Themes"
+            }
+        }
+
+        val themeSelection = ThemeSelectionDropdown()
+
+        val uiScaleSelector = JSpinner(
+            SpinnerNumberModel(Kindling.session.uiScaleFactor, 1.0, 2.0, 0.1)
+        )
+
+        init {
+            border = BorderFactory.createTitledBorder("Appearance/UI")
+
+            val themeLabel = JLabel("Theme")
+            val uiScaleLabel = JLabel("UI Scale (Requires restart)")
+
+            add(themeLabel, "grow")
+            add(lightDarkToggle)
+            add(themeSelection, "grow, wrap")
+
+            add(uiScaleLabel)
+            add(uiScaleSelector)
+        }
+
+        inner class ThemeSelectionDropdown private constructor(
+            private val lightThemes: Array<Kindling.Theme>,
+            private val darkThemes: Array<Kindling.Theme>,
+        ) : JComboBox<Kindling.Theme>() {
+
+            constructor() : this(Kindling.Theme.lightThemes.toTypedArray(), Kindling.Theme.darkThemes.toTypedArray())
+
+            init {
+                val initialTheme = Kindling.session.theme
+                select(initialTheme)
+
+                configureCellRenderer { _, value, _, _, _ ->
+                    text = value?.name
+                }
+
+                lightDarkToggle.addItemListener { event ->
+                    model = if (event.stateChange == ItemEvent.SELECTED) {
+                        DefaultComboBoxModel(lightThemes)
+                    } else {
+                        DefaultComboBoxModel(darkThemes)
+                    }
+                    setLafForSelectedItem()
+                }
+
+                addActionListener {
+                    setLafForSelectedItem()
+                }
+
+                Kindling.addThemeChangeListener { newTheme ->
```

Comment:
Need to remove this listener which doesn't do anything. It had a purpose at one point.